design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+
+We considered representing column types in a more structured form (rather than plain strings) to accomodate complex/nested data types but felt this added complexity would not be helpful for most use cases. (This would also require a server-side change, so it's a fair bit of additional work.)
+
+For `getDecimal(...)` in the `Row` interface, rather than trying to parse the column type to extract the precision and scale, we will simply convert the value to a BigDecimal without explicitly specifying the precision and scale. We could also add an option for users to specify the scale and precision in the getter, if we think that would be useful.
+
+We could also add `getList(...)` and `getMap(...)` methods to the `Row` interface, but it's not clear to me how valuable this would be given that the nested data types would not be known.
+
+Tim also proposed switching from generic Java types (`Map<String, Object>`, `List<Object>`) to Vert.x types (`JsonObject`, `JsonArray`) as the latter is more type-safe and the client already has a dependency on Vert.x anyway. The downside, though, is that then apps that use the client would be required to depend on Vert.x as well.
+
+### Transient queries -- Non-streaming
+
+The `Client` interface will also provide the following methods for receiving the results of a transient query (push or pull) in a single batch (non-streaming),
+once the query has completed:
+```
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql);
+
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @param properties query properties.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql, Map<String, Object> properties);
+```
+
+For a query to "complete" could mean:
+* The query is a pull query
+* The query is a push query with a limit clause, and the limit has been reached
+* The query is a push query that has been terminated
+
+We may want to introduce a limit to the number of rows that may be returned from these `executeQuery()` methods,
+in order to decrease the likelihood of running out of memory.
+
+### Insert values
+
+A method to insert one row at a time:
+```
+  /**
+   * Insert a single row into the relevant stream/table.
+   *
+   * @param streamName name of stream/table.
+   * @param row the row to insert.
+   * @return a future that completes once the request has been processed.
+   */
+  CompletableFuture<Void> insertInto(String streamName, Map<String, Object> row);
+```
+
+A method to stream inserts (via the `/inserts-stream` endpoint):
+```
+  CompletableFuture<Publisher<InsertResponse>> streamInserts(String streamName, Publisher<List<Object>> insertsPublisher);
```

Comment:
I think this would need to be:
```
CompletableFuture<Publisher<InsertResponse>> streamInserts(String streamName, Publisher<Map<String, Object>> insertsPublisher);
```
As, for inserting, a row is represented as a map.